### PR TITLE
fix: [PL-38828]: Added RBAC check to list Connectors when specific selected.

### PR DESCRIPTION
**Pull request by Jimit09Gandhi**
## Describe your changes
Added RBAC check to list Connectors when specific selected.
All RBAC checks are kept at resource layer.

Tests:
Performed following tests for all 4 Connector APIs where changes are done:
list
List/v2
Stats
lisCCMK8s
- List connectors when access to all connectors in scope.
- List connectors when access to specific connectors selected in Resource group.

## Checklist
- [X] I've documented the changes in the PR description.
- [] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/48825)

### CODE REVIEW

File: 120-ng-manager/src/main/java/io/harness/ng/ConnectorServiceImpl.java [hunk 1]
```diff
@@ -756,29 +758,54 @@ public void resetHeartBeatTask(String accountId, String taskId) {
   @Override
   public Page<ConnectorResponseDTO> list(String accountIdentifier, ConnectorFilterPropertiesDTO filterProperties,
       String orgIdentifier, String projectIdentifier, String filterIdentifier, String searchTerm,
-      Boolean includeAllConnectorsAccessibleAtScope, Boolean getDistinctFromBranches, Pageable pageable) {
+      Boolean includeAllConnectorsAccessibleAtScope, Boolean getDistinctFromBranches, Pageable pageable,
+      List<String> connectorIds) {
```
Comment (meenakshi-harness) [inline 120-ng-manager/src/main/java/io/harness/ng/ConnectorServiceImpl.java:762]: Can we use filterProperties for sending connector Ids?

File: 120-ng-manager/src/main/java/io/harness/ng/ConnectorServiceImpl.java [hunk 2]
```diff
@@ -794,9 +821,10 @@ public void deleteBatch(
   public Page<CombineCcmK8sConnectorResponseDTO> listCcmK8S(String accountIdentifier,
       ConnectorFilterPropertiesDTO filterProperties, String orgIdentifier, String projectIdentifier,
       String filterIdentifier, String searchTerm, Boolean includeAllConnectorsAccessibleAtScope,
-      Boolean getDistinctFromBranches, Pageable pageable) {
+      Boolean getDistinctFromBranches, Pageable pageable, List<String> connectorIds) {
```
Comment (meenakshi-harness) [inline 120-ng-manager/src/main/java/io/harness/ng/ConnectorServiceImpl.java:824]: same

File: 440-connector-nextgen/src/main/java/io/harness/connector/apis/resource/ConnectorResource.java [hunk 3]
```diff
@@ -252,10 +258,21 @@ public ConnectorResource(@Named("connectorDecoratorService") ConnectorService co
...
+            Resource.of(ResourceTypes.CONNECTOR, null), VIEW_CONNECTOR_PERMISSION)) {
+      return ResponseDTO.newResponse(getNGPageResponse(connectorService.list(page, size, accountIdentifier,
+          orgIdentifier, projectIdentifier, searchTerm, type, category, sourceCategory, version, emptyList())));
+    }
+    Page<Connector> allConnectors = connectorService.listAll(page, size, accountIdentifier, orgIdentifier,
+        projectIdentifier, searchTerm, type, category, sourceCategory, version);
+
+    List<Connector> permittedConnectors = connectorRbacHelper.getPermitted(allConnectors.getContent());
+    if (isEmpty(permittedConnectors)) {
+      return ResponseDTO.newResponse(getNGPageResponse(Page.empty()));
```
Comment (meenakshi-harness) [inline 440-connector-nextgen/src/main/java/io/harness/connector/apis/resource/ConnectorResource.java:271]: We should throw an error here for user not having access to view connector.

File: 440-connector-nextgen/src/main/java/io/harness/connector/apis/resource/ConnectorResource.java [hunk 4]
```diff
@@ -292,15 +309,28 @@ public ConnectorResource(@Named("connectorDecoratorService") ConnectorService co
...
+          orgIdentifier, projectIdentifier, filterIdentifier, searchTerm, includeAllConnectorsAccessibleAtScope,
+          getDistinctFromBranches, getPageRequest(pageRequest), version, emptyList())));
+    }
+    Page<Connector> allConnectors = connectorService.listAll(accountIdentifier, connectorListFilter, orgIdentifier,
+        projectIdentifier, filterIdentifier, searchTerm, includeAllConnectorsAccessibleAtScope, getDistinctFromBranches,
+        getPageRequest(pageRequest), version);
+
+    List<Connector> permittedConnectors = connectorRbacHelper.getPermitted(allConnectors.getContent());
+    if (isEmpty(permittedConnectors)) {
+      return ResponseDTO.newResponse(getNGPageResponse(Page.empty()));
```
Comment (meenakshi-harness) [inline 440-connector-nextgen/src/main/java/io/harness/connector/apis/resource/ConnectorResource.java:328]: same as above

File: 440-connector-nextgen/src/main/java/io/harness/connector/apis/resource/ConnectorResource.java [hunk 5]
```diff
@@ -252,10 +258,21 @@ public ConnectorResource(@Named("connectorDecoratorService") ConnectorService co
               + "CLOUD_PROVIDER, SECRET_MANAGER, CLOUD_COST, ARTIFACTORY, CODE_REPO,  "
               + "MONITORING and TICKETING") @QueryParam(SOURCE_CATEGORY_KEY) ConnectorCategory sourceCategory,
       @QueryParam("version") String version, @BeanParam GitEntityFindInfoDTO gitEntityBasicInfo) {
-    accessControlClient.checkForAccessOrThrow(ResourceScope.of(accountIdentifier, orgIdentifier, projectIdentifier),
-        Resource.of(ResourceTypes.CONNECTOR, null), VIEW_CONNECTOR_PERMISSION);
+    if (accessControlClient.hasAccess(ResourceScope.of(accountIdentifier, orgIdentifier, projectIdentifier),
```
Comment (meenakshi-harness) [inline 440-connector-nextgen/src/main/java/io/harness/connector/apis/resource/ConnectorResource.java:261]: This API is deprecated, can we check if its not used in UI, we can skip the changes here

File: 440-connector-nextgen/src/main/java/io/harness/connector/apis/resource/ConnectorResource.java [hunk 6]
```diff
@@ -561,10 +604,20 @@ public ConnectorResource(@Named("connectorDecoratorService") ConnectorService co
...
+            Resource.of(ResourceTypes.CONNECTOR, null), VIEW_CONNECTOR_PERMISSION)) {
+      return ResponseDTO.newResponse(
+          connectorService.getConnectorStatistics(accountIdentifier, orgIdentifier, projectIdentifier, emptyList()));
+    }
+    Page<Connector> allConnectors = connectorService.listAll(accountIdentifier, orgIdentifier, projectIdentifier);
+    List<Connector> permittedConnectors = connectorRbacHelper.getPermitted(allConnectors.getContent());
+    if (isEmpty(permittedConnectors)) {
+      return ResponseDTO.newResponse(
+          ConnectorStatistics.builder().statusStats(emptyList()).typeStats(emptyList()).build());
```
Comment (meenakshi-harness) [inline 440-connector-nextgen/src/main/java/io/harness/connector/apis/resource/ConnectorResource.java:616]: same here

**Review by meenakshi-harness — CHANGES_REQUESTED** [review]

File: 440-connector-nextgen/src/main/java/io/harness/connector/helper/ConnectorRbacHelper.java [hunk 7]
```diff
@@ -103,4 +116,55 @@ private void validateTheScopeOfTheSecret(SecretRefData secretRefData, NGAccess n
       }
     }
   }
+
+  public List<Connector> getPermitted(List<Connector> connectors) {
+    if (isEmpty(connectors)) {
+      return Collections.emptyList();
+    }
+    Map<EntityScopeInfo, List<Connector>> connectorsMap =
```
Comment (nishantsaini20) [inline 440-connector-nextgen/src/main/java/io/harness/connector/helper/ConnectorRbacHelper.java:124]: Does `EntityScopeInfo` has equals and hashcode methods correctly implemented?

File: 440-connector-nextgen/src/main/java/io/harness/connector/impl/ConnectorFilterServiceImpl.java [hunk 8]
```diff
@@ -89,6 +90,9 @@ public Criteria createCriteriaFromConnectorListQueryParams(String accountIdentif
       criteria.and(ConnectorKeys.orgIdentifier).is(orgIdentifier);
       criteria.and(ConnectorKeys.projectIdentifier).is(projectIdentifier);
     }
+    if (!isEmpty(connectorIds)) {
+      criteria.and(ConnectorKeys.id).in(connectorIds);
```
Comment (nishantsaini20) [inline 440-connector-nextgen/src/main/java/io/harness/connector/impl/ConnectorFilterServiceImpl.java:94]: Is this intended for `_id` field? Please verify when using `ConnectorKeys.id` correct index is getting used or not.

**Review by Jimit09Gandhi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48825)* status: 28 file(s) to review. 1 change requests. 6. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/ConnectorServiceImpl.java` line 762 at r2](https://reviewable.io/reviews/harness/harness-core/48825#gh-1223846349:-NXor-EW3SoR-dAjgcUr:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/b3dd4e5280f6996176deec0bdf6f2fd7b649aad1/120-ng-manager/src/main/java/io/harness/ng/ConnectorServiceImpl.java#L762)):*
<details><summary><i>Previously, meenakshi-harness (Meenakshi Raikwar) wrote…</i></summary><blockquote>

Can we use filterProperties for sending connector Ids?
</blockquote></details>

Done.
___
*[`120-ng-manager/src/main/java/io/harness/ng/ConnectorServiceImpl.java` line 824 at r2](https://reviewable.io/reviews/harness/harness-core/48825#gh-1223847527:-NXor-a9-c_sRLNHPIZA:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/b3dd4e5280f6996176deec0bdf6f2fd7b649aad1/120-ng-manager/src/main/java/io/harness/ng/ConnectorServiceImpl.java#L824)):*
<details><summary><i>Previously, meenakshi-harness (Meenakshi Raikwar) wrote…</i></summary><blockquote>

same
</blockquote></details>

Done.
___
*[`440-connector-nextgen/src/main/java/io/harness/connector/apis/resource/ConnectorResource.java` line 261 at r2](https://reviewable.io/reviews/harness/harness-core/48825#gh-1223850354:-NXor0fH8fMlV6d14uAR:b-ho0ub9) ([raw file](https://github.com/harness/harness-core/blob/b3dd4e5280f6996176deec0bdf6f2fd7b649aad1/440-connector-nextgen/src/main/java/io/harness/connector/apis/resource/ConnectorResource.java#L261)):*
<details><summary><i>Previously, meenakshi-harness (Meenakshi Raikwar) wrote…</i></summary><blockquote>

This API is deprecated, can we check if its not used in UI, we can skip the changes here
</blockquote></details>

This API is invoked from UI.
___
*[`440-connector-nextgen/src/main/java/io/harness/connector/apis/resource/ConnectorResource.java` line 271 at r2](https://reviewable.io/reviews/harness/harness-core/48825#gh-1223849788:-NXorB4Q1dFDVsopp4rP:b-4lyrh4) ([raw file](https://github.com/harness/harness-core/blob/b3dd4e5280f6996176deec0bdf6f2fd7b649aad1/440-connector-nextgen/src/main/java/io/harness/connector/apis/resource/ConnectorResource.java#L271)):*
<details><summary><i>Previously, meenakshi-harness (Meenakshi Raikwar) wrote…</i></summary><blockquote>

We should throw an error here for user not having access to view connector.
</blockquote></details>

Done. Implemented in getPermitted.
___
*[`440-connector-nextgen/src/main/java/io/harness/connector/apis/resource/ConnectorResource.java` line 328 at r2](https://reviewable.io/reviews/harness/harness-core/48825#gh-1223850074:-NXorKYY1eKH997At-rI:b-4lyrh4) ([raw file](https://github.com/harness/harness-core/blob/b3dd4e5280f6996176deec0bdf6f2fd7b649aad1/440-connector-nextgen/src/main/java/io/harness/connector/apis/resource/ConnectorResource.java#L328)):*
<details><summary><i>Previously, meenakshi-harness (Meenakshi Raikwar) wrote…</i></summary><blockquote>

same as above
</blockquote></details>

Done. Implemented in getPermitted.
___
*[`440-connector-nextgen/src/main/java/io/harness/connector/apis/resource/ConnectorResource.java` line 616 at r2](https://reviewable.io/reviews/harness/harness-core/48825#gh-1223851153:-NXorLOC3u-73Rtf0YA-:b-4lyrh4) ([raw file](https://github.com/harness/harness-core/blob/b3dd4e5280f6996176deec0bdf6f2fd7b649aad1/440-connector-nextgen/src/main/java/io/harness/connector/apis/resource/ConnectorResource.java#L616)):*
<details><summary><i>Previously, meenakshi-harness (Meenakshi Raikwar) wrote…</i></summary><blockquote>

same here
</blockquote></details>

Done. Implemented in getPermitted.

**Comment by Jimit09Gandhi** [comment]
trigger uts

**Comment by Jimit09Gandhi** [comment]
trigger uts

**Review by Jimit09Gandhi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48825)* status: 30 file(s) to review. 1 change requests. 7. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`440-connector-nextgen/src/main/java/io/harness/connector/impl/ConnectorFilterServiceImpl.java` line 94 at r2](https://reviewable.io/reviews/harness/harness-core/48825#gh-1226193161:-NYIcO4xDb7kEbz6bUki:bivu6v8) ([raw file](https://github.com/harness/harness-core/blob/b3dd4e5280f6996176deec0bdf6f2fd7b649aad1/440-connector-nextgen/src/main/java/io/harness/connector/impl/ConnectorFilterServiceImpl.java#L94)):*
<details><summary><i>Previously, nishantsaini20 (Nishant Saini) wrote…</i></summary><blockquote>

Is this intended for `_id` field? Please verify when using `ConnectorKeys.id` correct index is getting used or not. 
</blockquote></details>

yes, checked it will use existing indexes or index on _id field.

**Review by Jimit09Gandhi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48825)* status: 30 file(s) to review. 1 change requests. 8. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`440-connector-nextgen/src/main/java/io/harness/connector/helper/ConnectorRbacHelper.java` line 124 at r2](https://reviewable.io/reviews/harness/harness-core/48825#gh-1226188380:-NYJ-Je5CRQeB3HM6cPQ:b-3vkcdu) ([raw file](https://github.com/harness/harness-core/blob/b3dd4e5280f6996176deec0bdf6f2fd7b649aad1/440-connector-nextgen/src/main/java/io/harness/connector/helper/ConnectorRbacHelper.java#L124)):*
<details><summary><i>Previously, nishantsaini20 (Nishant Saini) wrote…</i></summary><blockquote>

Does `EntityScopeInfo` has equals and hashcode methods correctly implemented?
</blockquote></details>

Yes `EntityScopeInfo` is annotated with @Value and @Value internally uses @EqualsAndHashCode.
For details refer: https://projectlombok.org/features/Value

**Review by nishantsaini20 — APPROVED** [review]
> Reviewed 1 of 11 files at r1, 1 of 26 files at r2, 2 of 17 files at r3, 17 of 25 files at r4, 3 of 3 files at r5, 3 of 3 files at r6, 3 of 3 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48825)* status: 30 file(s) to review. 1 change requests. 6. unresolved discussions. 1 of 2 approvals obtained.

**Pull request merged**